pkg/logql/log/parser.go

```diff
@@ -26,67 +28,169 @@ var (
 	errMissingCapture = errors.New("at least one named capture must be supplied")
 )
 
-func addLabel(lbs *LabelsBuilder, key, value string) {
-	key = sanitizeKey(key)
-	if lbs.BaseHas(key) {
-		key = fmt.Sprintf("%s%s", key, duplicateSuffix)
+type JSONParser struct {
+	buf []byte // buffer used to build json keys
+	lbs *LabelsBuilder
+}
+
+// NewJSONParser creates a log stage that can parse a json log line and add properties as labels.
+func NewJSONParser() *JSONParser {
+	return &JSONParser{
+		buf: make([]byte, 0, 1024),
 	}
-	lbs.Set(key, value)
 }
 
-func sanitizeKey(key string) string {
-	if len(key) == 0 {
-		return key
+func (j *JSONParser) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+	it := jsoniter.ConfigFastest.BorrowIterator(line)
+	defer jsoniter.ConfigFastest.ReturnIterator(it)
+
+	// reset the state.
+	j.buf = j.buf[:0]
+	j.lbs = lbs
+
+	if err := j.readObject(it); err != nil {
+		lbs.SetErr(errJSON)
+		return line, true
+	}
+	return line, true
+}
+
+func (j *JSONParser) readObject(it *jsoniter.Iterator) error {
+	// we only care about object and values.
+	if nextType := it.WhatIsNext(); nextType != jsoniter.ObjectValue {
+		return errors.New("not a json object")
```

Comment:
WDYT about `"expecting json object, got %s"`?

Reply:
Yep the error is swallowed right after, I'll still make the change, not sure I want to log it yet, this is happening for every lines.